src/engine/V3Eval.v3

```diff
@@ -835,6 +847,31 @@ def do_vv_v_x16(a: (u64, u64), b: (u64, u64), f: (u8, u8) -> u8) -> (u64, u64) {
 	}
 		return (low, high);
 }
+// Helper to perform v128 extensions.
+// T: type to be converted
+// Tw: wider type of T, i.e., the extended type
+// Uw: unsigned type of with the same width as Tw
+// S: source type of v128.{low/high}, always u64 but need to be explicit put here for type inference
+def V128_INT_EXTEND<T, Tw, Uw, S>(a: (u64, u64), low_half: bool, dest_lanes: byte,
+convert: (S) -> T, extend: (T) -> Tw, convert_to_u: (Tw) -> Uw) -> (u64, u64) {
+	def half_lanes = dest_lanes >> 1;
+	var src = if(low_half, a.0, a.1);	// Choose low or high half of source lanes.
+	var low: u64 = 0u, high: u64 = 0u;
+	var shift = byte.view(64 / dest_lanes);	// Compute source shift amount, i.e., the width of the original lane.
+	def mask: u64 = 0xffffffffffffffffuL >> byte.view(64 - shift);
+	for (i < dest_lanes) {
```

Comment:
You could maybe split this into two loops? AFAICT the first `half_lane` iterations do one thing and the other half do another thing. You could even split this into two methods?

Reply:
Thanks! I was thinking about that in the same way. Pushed a new change.